# Movies Tonight Part 3: SQL DML

In [3]:
%load_ext sql
%sql sqlite:///Movies.db

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [7]:
%%sql
SELECT *
FROM DATASET
LIMIT 10;

 * sqlite:///Movies.db
Done.


[('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Austin Pendleton'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Bebe Neuwirth'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Dianne Wiest'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Eli Wallach'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Kenny Kerr'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Lainie Kazan'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Tim Daly'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Whoopi Goldberg'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'D', 'Donald Petrie'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '7:20 PM', 'PG-13', 'A', 'Austin Pendleton')]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [3]:
%%sql
INSERT INTO MOVIE (MTitle, Rating)
    SELECT DISTINCT MTitle, Rating
    FROM DATASET;
    

 * sqlite:///Movies.db
23 rows affected.


[]

In [4]:
%%sql
INSERT INTO THEATER (TName, Location, Phone)
    SELECT DISTINCT TName, Location, Phone
    FROM DATASET;

 * sqlite:///Movies.db
9 rows affected.


[]

In [5]:
%%sql
INSERT INTO ARTIST (Name)
    SELECT DISTINCT CName
    FROM DATASET;

 * sqlite:///Movies.db
153 rows affected.


[]

In [6]:
%%sql
INSERT INTO CREDIT (MID, AID, CCode)
SELECT DISTINCT MID, AID, CCode
FROM DATASET
    JOIN MOVIE ON (MOVIE.MTitle = DATASET.MTitle)    
    JOIN ARTIST ON (ARTIST.Name = DATASET.CName);

 * sqlite:///Movies.db
162 rows affected.


[]

In [7]:
%%sql
INSERT INTO SHOW (MID, TID, STime)
SELECT DISTINCT MID,TID, ShowTime
FROM DATASET
    JOIN MOVIE ON (DATASET.MTitle = MOVIE.MTitle)
    JOIN THEATER ON (DATASET.TName = THEATER.TName);

 * sqlite:///Movies.db
131 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [ ]:
YOUR CODE HERE

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [ ]:
YOUR CODE HERE

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".